Add wait mechanism to verify_volume_deleted_in_backend() in helpers.py

In tests.helpers `verify_volume_deleted_in_backend` should wait for backend image deletion to complete.
We have seen some tier4 tests failing intermittently with the following similar error
```
AssertionError: Volume associated with PVC pvc-test-44573f761a594ba080b221884afe4831 still exists in backend assert False
```

One such run [RUN ID: 1588554726](https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/7221/testReport/)

When checked in the cluster afterwards the backend images from the testrun were already deleted

PVs after the test run:
```
$ oc get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                                           STORAGECLASS                  REASON   AGE
local-pv-3e8964d3                          1862Gi     RWO            Delete           Bound    openshift-storage/ocs-deviceset-2-0-679j5                       localblock                             3d16h
local-pv-414755e0                          1862Gi     RWO            Delete           Bound    openshift-storage/ocs-deviceset-1-0-wr67z                       localblock                             3d16h
local-pv-d9c5cbd6                          1862Gi     RWO            Delete           Bound    openshift-storage/ocs-deviceset-0-0-pp9w8                       localblock                             3d16h
pvc-0037ef59-9e83-47d8-8ce2-714941a431ea   40Gi       RWO            Delete           Bound    openshift-monitoring/prometheus-k8s-db-prometheus-k8s-1         ocs-storagecluster-ceph-rbd            3d16h
pvc-37f82ba6-d091-4ed5-8d5c-911b58b9780b   40Gi       RWO            Delete           Bound    openshift-monitoring/alertmanager-main-db-alertmanager-main-0   ocs-storagecluster-ceph-rbd            3d16h
pvc-6fb648c1-f7c4-4116-8d7a-1df8c8176277   40Gi       RWO            Delete           Bound    openshift-monitoring/alertmanager-main-db-alertmanager-main-1   ocs-storagecluster-ceph-rbd            3d16h
pvc-8ef33313-c729-43ea-ab7a-fc10fd279445   40Gi       RWO            Delete           Bound    openshift-monitoring/alertmanager-main-db-alertmanager-main-2   ocs-storagecluster-ceph-rbd            3d16h
pvc-9253274e-22ff-4eb0-85a9-8c40f5862092   50Gi       RWO            Delete           Bound    openshift-storage/db-noobaa-db-0                                ocs-storagecluster-ceph-rbd            3d16h
pvc-991578d1-3ecb-44a7-84d5-c7ffc5e64818   40Gi       RWO            Delete           Bound    openshift-monitoring/prometheus-k8s-db-prometheus-k8s-0         ocs-storagecluster-ceph-rbd            3d16h
pvc-d3ef972f-c9c1-470f-94a8-8f2a2795647d   100Gi      RWX            Delete           Bound    openshift-image-registry/registry-cephfs-rwx-pvc                ocs-storagecluster-cephfs              3d16h
```

From Ceph Toolbox pod:

```
# rbd ls -l -p ocs-storagecluster-cephblockpool
NAME                                         SIZE   PARENT FMT PROT LOCK 
csi-vol-563949fb-8bc3-11ea-b0e8-0a580a830012 50 GiB          2           
csi-vol-c42235b0-8bc3-11ea-b0e8-0a580a830012 40 GiB          2           
csi-vol-c428a548-8bc3-11ea-b0e8-0a580a830012 40 GiB          2           
csi-vol-c42fecf9-8bc3-11ea-b0e8-0a580a830012 40 GiB          2           
csi-vol-c5a9ba57-8bc3-11ea-b0e8-0a580a830012 40 GiB          2           
csi-vol-c5b08b22-8bc3-11ea-b0e8-0a580a830012 40 GiB          2
```

--- a/tests/helpers.py
+++ b/tests/helpers.py
@@ -1354,7 +1354,8 @@ def is_volume_present_in_backend(interface, image_uuid, pool_name=None):
         pool_name (str): Name of the rbd-pool if interface is CephBlockPool
 
     Returns:
-        bool: True if volume is present. False if volume is not present.
+        bool: True if volume is present and False if volume is not present
+
     """
     ct_pod = pod.get_ceph_tools_pod()
     if interface == constants.CEPHBLOCKPOOL:

--- a/tests/helpers.py
+++ b/tests/helpers.py
@@ -1426,6 +1426,10 @@ def verify_volume_deleted_in_backend(
             f"Volume corresponding to uuid {image_uuid} is not deleted "
             f"in backend"
         )
+        # Log 'ceph progress' and 'ceph rbd task list' for debugging purpose
+        ct_pod = pod.get_ceph_tools_pod()
+        ct_pod.exec_ceph_cmd('ceph progress')
+        ct_pod.exec_ceph_cmd('ceph rbd task list')
         return False
 
 

--- a/tests/helpers.py
+++ b/tests/helpers.py
@@ -1338,9 +1338,9 @@ def change_default_storageclass(scname):
     return True
 
 
-def verify_volume_deleted_in_backend(interface, image_uuid, pool_name=None):
+def is_volume_present_in_backend(interface, image_uuid, pool_name=None):
     """
-    Verify that Image/Subvolume is not present in the backend.
+    Check whether Image/Subvolume is present in the backend.
 
     Args:
         interface (str): The interface backed the PVC
@@ -1354,7 +1354,7 @@ def verify_volume_deleted_in_backend(interface, image_uuid, pool_name=None):
         pool_name (str): Name of the rbd-pool if interface is CephBlockPool
 
     Returns:
-        bool: True if volume is not present. False if volume is present
+        bool: True if volume is present. False if volume is not present.
     """
     ct_pod = pod.get_ceph_tools_pod()
     if interface == constants.CEPHBLOCKPOOL:
@@ -1369,17 +1369,63 @@ def verify_volume_deleted_in_backend(interface, image_uuid, pool_name=None):
 
     try:
         ct_pod.exec_ceph_cmd(ceph_cmd=cmd, format='json')
-        return False
+        logger.info(
+            f"Verified: Volume corresponding to uuid {image_uuid} exists "
+            f"in backend"
+        )
+        return True
     except CommandFailed as ecf:
         assert valid_error in str(ecf), (
-            f"Error occurred while verifying volume is deleted in backend: "
+            f"Error occurred while verifying volume is present in backend: "
             f"{str(ecf)} ImageUUID: {image_uuid}. Interface type: {interface}"
         )
-    logger.info(
-        f"Verified: Volume corresponding to uuid {image_uuid} is deleted "
-        f"in backend"
-    )
-    return True
+        logger.info(
+            f"Volume corresponding to uuid {image_uuid} does not exist "
+            f"in backend"
+        )
+        return False
+
+
+def verify_volume_deleted_in_backend(
+    interface, image_uuid, pool_name=None, timeout=180
+):
+    """
+    Ensure that Image/Subvolume is deleted in the backend.
+
+    Args:
+        interface (str): The interface backed the PVC
+        image_uuid (str): Part of VolID which represents
+            corresponding image/subvolume in backend
+            eg: oc get pv/<volumeName> -o jsonpath='{.spec.csi.volumeHandle}'
+                Output is the CSI generated VolID and looks like:
+                '0001-000c-rook-cluster-0000000000000001-
+                f301898c-a192-11e9-852a-1eeeb6975c91' where
+                image_uuid is 'f301898c-a192-11e9-852a-1eeeb6975c91'
+        pool_name (str): Name of the rbd-pool if interface is CephBlockPool
+        timeout (int): Wait time for the volume to be deleted.
+
+    Returns:
+        bool: True if volume is deleted before timeout.
+            False if volume is not deleted.
+    """
+    try:
+        for ret in TimeoutSampler(
+            timeout, 2, is_volume_present_in_backend, interface=interface,
+            image_uuid=image_uuid, pool_name=pool_name
+        ):
+            if not ret:
+                break
+        logger.info(
+            f"Verified: Volume corresponding to uuid {image_uuid} is deleted "
+            f"in backend"
+        )
+        return True
+    except TimeoutExpiredError:
+        logger.error(
+            f"Volume corresponding to uuid {image_uuid} is not deleted "
+            f"in backend"
+        )
+        return False
 
 
 def create_serviceaccount(namespace):

--- a/tests/manage/pv_services/test_change_reclaim_policy_of_pv.py
+++ b/tests/manage/pv_services/test_change_reclaim_policy_of_pv.py
@@ -297,17 +297,7 @@ class TestChangeReclaimPolicyOfPv(ManageTest):
         # Verify PV using ceph toolbox. Wait for Image/Subvolume to be deleted.
         pool_name = default_ceph_block_pool() if interface == constants.CEPHBLOCKPOOL else None
         for pvc_name, uuid in pvc_uuid_map.items():
-            try:
-                for ret in TimeoutSampler(
-                    180, 2, verify_volume_deleted_in_backend,
-                    interface=interface, image_uuid=uuid, pool_name=pool_name
-                ):
-                    if ret:
-                        break
-            except TimeoutExpiredError as err:
-                err.message = (
-                    f"{err.message}- Volume associated with PVC {pvc_name} "
-                    f"still exists in backend"
-                )
-                raise
+            assert verify_volume_deleted_in_backend(
+                interface=interface, image_uuid=uuid, pool_name=pool_name
+            ), f"Volume associated with PVC {pvc_name} still exists in backend"
         log.info("Verified: Image/Subvolume removed from backend.")

--- a/tests/manage/pv_services/test_change_reclaim_policy_of_pv.py
+++ b/tests/manage/pv_services/test_change_reclaim_policy_of_pv.py
@@ -9,7 +9,6 @@ from tests.helpers import (
     wait_for_resource_state, verify_volume_deleted_in_backend,
     default_ceph_block_pool
 )
-from ocs_ci.ocs.exceptions import TimeoutExpiredError
 
 log = logging.getLogger(__name__)
 

